### PR TITLE
PETSc fix + SLEPc fix + CI

### PR DESCRIPTION
**Pull request by dschwoerer**
Currently Fedora CI is failing, as new PETSc has landed.

### CODE REVIEW

File: src/solver/impls/slepc/slepc.cxx [hunk 1]
```diff
@@ -162,11 +162,17 @@ SlepcSolver::SlepcSolver(Options* options) {
              .withDefault(0);
 
   tol = options_ref["tol"].doc("SLEPc tolerance").withDefault(1.0e-6);
-  maxIt = options_ref["maxIt"].doc("Maximum iterations").withDefault(PETSC_DEFAULT);
+  maxIt = options_ref["maxIt"].doc("Maximum iterations").withDefault(-1);
+  if (maxIt == -1) {
+    maxIt = PETSC_DEFAULT;
+  }
```
Comment (ZedThree) [inline src/solver/impls/slepc/slepc.cxx:168]: PETSc is so incredibly frustrating sometimes. I got grumpy and asked them to revert this change. This is obviously fine as a workaround from our side, but I'm annoyed we have to do this.

**Comment by ZedThree** [comment]
Failing tests are:

- unit tests, with an error like this:

```
/lib/x86_64-linux-gnu/libstdc++.so.6(_Znwm+0x19)[0x7f00b3df4b39]
[fv-az953-5:30138] [ 8] /home/runner/work/BOUT-dev/BOUT-dev/build/lib/libgtest.so.1.10.0(+0x2531d)[0x7f00b583931d]
[fv-az953-5:30138] [ 9] /home/runner/work/BOUT-dev/BOUT-dev/build/lib/libgtest.so.1.10.0(_ZN7testing10TestResult17AddTestPartResultERKNS_14TestPartResultE+0x64)[0x7f00b5843c84]
[fv-az953-5:30138] [10] /home/runner/work/BOUT-dev/BOUT-dev/build/lib/libgtest.so.1.10.0(_ZN7testing8internal35DefaultGlobalTestPartResultReporter20ReportTestPartResultERKNS_14TestPartResultE+0x24)[0x7f00b5843d04]
[fv-az953-5:30138] [11] /home/runner/work/BOUT-dev/BOUT-dev/build/lib/libgtest.so.1.10.0(_ZN7testing8UnitTest17AddTestPartResultENS_14TestPartResult4TypeEPKciRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESC_+0x28c)[0x7f00b584f07c]
[fv-az953-5:30138] [12] /home/runner/work/BOUT-dev/BOUT-dev/build/lib/libgtest.so.1.10.0(_ZNK7testing8internal12AssertHelperaSERKNS_7MessageE+0x78)[0x7f00b584f228]
[fv-az953-5:30138] [13] /home/runner/work/BOUT-dev/BOUT-dev/tests/unit/invert/laplace/test_laplace_petsc3damg.cxx:326: Failure
The difference between expected[i] and actual[i] is 2.9564733378985297e-07, which exceeds tol, where
expected[i] evaluates to 0.0025689451062514425,
actual[i] evaluates to 0.0025686494589176527, and
tol evaluates to 1e-08.
```

And 

- test-laplace-petsc3d (Failed)
- test-slepc-solver (Failed)

both with errors like:

```
[0]PETSC ERROR: Object is in wrong state
[0]PETSC ERROR: Not for unassembled vector, did you call VecAssemblyBegin()/VecAssemblyEnd()?
[0]PETSC ERROR: WARNING! There are option(s) set that were not used! Could be the program crashed before they were used or a spelling mistake, etc!
[0]PETSC ERROR:   Option left: name:-boutpetsclib_laplacemg_levels_ksp_max_it value: 3 source: code
[0]PETSC ERROR:   Option left: name:-d value: data_circular_core source: command line
```

Are we not assembling something correctly?

---

Also lots of warnings of:

```
PETSC_NULL is deprecated, use PETSC_NULLPTR instead (since version 3.19)
```

I don't understand why they didn't just change the definition of `PETSC_NULL`, they'll happily change everything else.

Maybe let's just replace our uses of `PETSC_NULL` with `nullptr`?

**Comment by dschwoerer** [comment]
So we fixed (at least to the extend that the tests pass) for 3.18, and now 3.19 broke it again :-1: 

Maybe we should try to run against PETSc devel to catch breaking changes early, and maybe have a chance to weigh in?

**Comment by ZedThree** [comment]
That sounds like a good idea

**Comment by ZedThree** [comment]
The reason the Laplace PETSc3D test is failing is due to a change in default `PCGAMG` in 3.17:

> Change [PCGAMG](https://petsc.org/release/manualpages/PC/PCGAMG/) default to use [PCJACOBI](https://petsc.org/release/manualpages/PC/PCJACOBI/) smoothing instead of [PCSOR](https://petsc.org/release/manualpages/PC/PCSOR/). This also allows the default configuration to use GPUs effectively, and to deliver equivalent convergence. For the old default, use -mg_levels_pc_type sor.

Use SOR does fix that one test at least.

Question is now what do we do? Do we always use SOR? It's not clear how to do this programmatically, except through setting string options. JACOBI doesn't seem to have many options, and none of them fix the test.

`laplace:petsc:mg_levels_ksp_max_it=4` does work but slows it down by factor 2x

**Comment by ZedThree** [comment]
I'm looking at fixes for these issues. I'm going to rebase to go into master in anticipation of new release soonish
